Finalize pwfilesupport

varet80 opened the pull request:
New Feature: Create a Passwd file using Databags

    - Generate UserDB
     - Source from databags
     - Compare users with the current if exist

    - Kitchen
     - Change Exceptions for centos 7.2 (fall into some bugs)
     - Add databags support
     - Add databags with two users

    - User Recipe
     - Change priority on dovecot user creation (group goes after user)

    - New AttribugesAttributes
     - Databag name attribute is  `databag_name`
     - Databag item for userdb `databag_users_item`

stromp committed:
now with reverse verification

stromp committed:
now with reverse verification

stromp committed:
now with reverse verification

stromp committed:
now with reverse verification

stromp committed:
new attrs

stromp committed:
new attrs

stromp committed:
cleaning

stromp committed:
cleaning

stromp committed:
cleaning

stromp committed:
cleaning

stromp committed:
adding template

stromp committed:
adding template

stromp committed:
cleaning up

stromp committed:
cleaning up

stromp committed:
beautify

stromp committed:
beautify

stromp committed:
updated metadata and readme

stromp committed:
updated metadata and readme

stromp committed:
swap user and group

stromp committed:
swap user and group

stromp committed:
Merge branch 'adding_pwfilesupport' of github.com:schubergphilis/dovecot-cookbook into adding_pwfilesupport

Commit:
Merge pull request #1 from schubergphilis/adding_pwfilesupport

now with reverse verification

Commit:
Version 2.4.1001

stromp committed:
Merge branch 'adding_pwfilesupport' of github.com:schubergphilis/dovecot-cookbook into adding_pwfilesupport

Conflicts:
	attributes/conf_files.rb

stromp committed:
Merge pull request #2 from schubergphilis/adding_pwfilesupport

Adding pwfilesupport

varet80 committed:
Merge remote-tracking branch 'upstream/master'

varet80 committed:
New Feature: Create a Passwd file using Databags

- Generate UserDB
 - Source from databags
 - Compare users with the current if exist

- Kitchen
 - Change Exceptions for centos 7.2 (fall into some bugs)
 - Add databags support
 - Add databags with two users

- User Recipe
 - Change priority on dovecot user creation (group goes after user)

- Rename Attributes
 - Databag attribute is now `databag_name` (more generic)
 - Databag item for users is now `databag_users_item`

varet80 committed:
Merge remote-tracking branch 'upstream/master' into finalize_pwfilesupport

billiaz commented:
@zuazo I might need your help with the unit tests & Integration tests, though the system generates properly the file on integration tests.

Also I want to see an alternative (next version) of using custom resource for the password file.

varet80 committed:
Unit testing
 - Add the missing replication file under unit testing
 - Add unit test for the file /etc/dovecot/password creation

Password File creation
- Change logic to handle issues like:
 - Missing /etc/dovecot/passwd file
 - Missing users under this file

zuazo commented:
OK, Thanks @billiaz!! Give me a couple of days to review this.

Regarding the tests, I think the simplest thing would be to do some [integration tests](https://github.com/zuazo/dovecot-cookbook/blob/master/TESTING.md#integration-tests-in-vagrant) in the testing cookbook, maybe create a new recipe there, and adding some [Serverspec resources to check the generated file](http://serverspec.org/resource_types.html#file) under [*test/integration*](https://github.com/zuazo/dovecot-cookbook/tree/master/test/integration). The steps:

- [ ] Create a new recipe: `dovecot_test::create_pwfile`. This recipe will include your `dovecot::create_pwfile` recipe.
- [ ] Create a data bag example to test it in *test/data_bags*.
- [ ] Add that recipe to a new suite in the *.kitchen.yml*.
- [ ] Add integration tests that check the generated pwfile in *test/integration/pwfile/serverspec* (or whatever the name of the suite is).
- [ ] Test the new suite at least in one platform on *.travis.yml*.

Tell me if you run into any errors/problems. If you can not complete them, send me what you have and I already try to do it myself.

Another alternative would be to add ChefSpec tests, but I think in this specific case that could be trickier due to the ruby block.

billiaz commented:
Hey! thanks for the support
1. I have already done the new recipe
2. databags too
3. I need to add the recipe on a new suite 
4. test the new suite
5. Move the block under custom resource (too big under rubocop)
redoo 2-4

varet80 committed:
Feature: Userdb file support (like passwd)

Refactoring User Creation
 - Use library with small functions
 - Add new simpler way of managing the changes
 - Add documentation and examples

Tests
 - Add unit testing
 - New kitchen suite
 - Server specs for the new kitchen suite

Rubocop:
 - Make lines wider

zuazo reviewed (changes requested):
OK. Thanks @billiaz! I added some comments to your code.

Don't worry too much about Custom Resources, that may be too much to be included in this PR. We can address it in a future issue/PR anyway.

billiaz commented on recipes/user.rb:
```diff
@@ -32,6 +32,14 @@
+else
+  user node['dovecot']['conf']['default_login_user']
+  group node['dovecot']['conf']['default_login_user']
+end
```
I have to check this, it was inherited.

billiaz replied:
legacy from the previous pull request, cleaning it and testing it

billiaz commented on recipes/create_pwfile.rb:
```diff
@@ -0,0 +1,95 @@
+#
+extend Chef::Mixin::ShellOut
+
+include_recipe 'dovecot::from_package'
```
Will update it on the next commit, Thnx

billiaz commented on README.md:
```diff
@@ -465,6 +468,12 @@ Installs the required packages. Used by the default recipe if `node['dovecot']['
+
+Creates and configures a password file from local mailboxes based on a data bag.
+
+* `node['dovecot']['databag_users_item']`: The databag item to use (under the databag set)
```
is it better to have it:
`node['dovecot']['databag_name']['users_item']` ?

billiaz commented on metadata.rb:
```diff
@@ -27,7 +27,7 @@
 EOH
 long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
-version '3.2.0' # WiP
+version '3.2.0001'
```
reverted

billiaz commented on recipes/create_pwfile.rb:
```diff
@@ -0,0 +1,95 @@
+                          [crypt, uid, gid, gecos, homedir, shell, extra_fields]
+                        else
+                          [crypt]
+                        end
```
refactored :) now inside libs
this accepts any type of userdb entry
there have been 2 ways:
user:pass (old)
current:
user:pass:uid:gid:gecos:homedir:shell:extra < some of these doesn't work
this is to force importing properly both, but now after I made some changes it should work better, with less complexity

varet80 committed:
Small fixes after Last commit

 - Fix Style syntax
 - Add travis checks for the new kitchen unit tests
 - Fix readme spaces under apache license

billiaz commented on README.md:
```diff
@@ -1076,9 +1129,9 @@ License and Author
-    
+
         http://www.apache.org/licenses/LICENSE-2.0
-    
```
fixing it on my next commit with the style changes :(
my lovely editor removes all white space

varet80 committed:
Kitchen and Travis modification:

Added kitchen and travis support
- Unfortunatelly not every test is completed as the config changes per
  dovecot version. (To be done on next versions)
- Newer versions require ssl (especially on latest Deb/Ub)

varet80 committed:
Fix attribute conditional to follow Rubocop rule Style/ConditionalAssignment

zuazo commented on attributes/conf_10_ssl.rb:
```diff
@@ -21,11 +21,10 @@
+default['dovecot']['conf']['ssl'] = \
+  if node['platform_family'] == 'suse' && node['platform_version'].to_i < 13
+    false
+  end
```
:+1:

zuazo commented on test/unit/recipes/create_pwfile_spec.rb:
```diff
@@ -0,0 +1,34 @@
+        .with_path('/etc/dovecot/password')
+        .with_source('password.erb')
+    end
+  end
```
This test [is failing](https://travis-ci.org/zuazo/dovecot-cookbook/jobs/208712070):

```
Failures:

  1) dovecot::create_pwfile when the module runs creates a file
     Failure/Error:
       expect(chef_run). to create_template('/etc/dovecot/password')
         .with_path('/etc/dovecot/password')
         .with_source('password.erb')
       expected "template[/etc/dovecot/password]" actions [] to include :create

     # ./test/unit/recipes/create_pwfile_spec.rb:29:in `block (3 levels) in <top (required)>'
```

That's because we need to run the code inside the `ruby_block` and force it to set `update_credentials` to `true`.

It is not easy to understand, but here you have a working version of ChefSpec tests for the `create_pwfile` recipe:

```ruby
describe 'dovecot::create_pwfile', order: :random do
  let(:chef_runner) { ChefSpec::SoloRunner.new }
  let(:chef_run) { chef_runner.converge(described_recipe) }
  let(:node) { chef_runner.node }

  it 'runs databag_to_dovecot_userdb ruby block' do
    expect(chef_run).to run_ruby_block('databag_to_dovecot_userdb')
  end

  describe 'inside ruby_block[databag_to_dovecot_userdb] resource' do
    let(:chef_runner) { ChefSpec::SoloRunner.new(step_into: %w(ruby_block)) }
    let(:data_bag_users) do
      {
        'users' => {
          'dilan' => 'password1234',
          'vassilis' => ['vassilis1234', nil, nil, nil, nil, nil, nil]
        }
      }
    end
    let(:pwfile) { instance_double('File', readlines: [], close: nil) }
    let(:encrypted_password) { 'encrypted_password' }
    before do
      # Stub ::File.open:
      allow(::File).to receive(:open).and_call_original
      allow(::File).to receive(:open).with('/etc/dovecot/password', any_args)
        .and_return(pwfile)
      # If passing a block with do...end to ::File.open:
      # allow(::File).to receive(:open).with('/etc/dovecot/password', any_args)
      #   .and_yield(pwfile)
  
      # Stub Data Bag:
      stub_data_bag_item('dovecot', 'users').and_return(data_bag_users)
  
      # Stub some commands excuted inside the ruby_block:
      # Using: https://github.com/zuazo/filesystem_resize-cookbook/blob/master/test/unit/support/stubs.rb
      stub_shell_out(
        '/usr/bin/doveadm pw -s MD5 -p                 password1234'
      ).and_return_stdout(encrypted_password)
      stub_shell_out(
        '/usr/bin/doveadm pw -s MD5 -p                 vassilis1234'
      ).and_return_stdout(encrypted_password)
    end

    it 'creates password file template' do
      expect(chef_run).to create_template('/etc/dovecot/password')
        .with_path('/etc/dovecot/password')
        .with_source('password.erb')
        .with_owner('dovecot')
        .with_group('dovecot')
        .with_mode('0640')
        # .with_sensitve(true)
    end
  end
end
```

This example uses the `stub_data_bag_item` function from [here](https://github.com/zuazo/filesystem_resize-cookbook/blob/master/test/unit/support/stubs.rb#L51-L57). You need to copy that file to the *test/unit/support* directory and include it in the [*test/unit/spec_helper.rb*](https://github.com/zuazo/dovecot-cookbook/blob/master/test/unit/spec_helper.rb#L26-L27) file.

If you change the `::File.open` call to use a block, as I have recommended you above, you need to uncomment the line with the `and_yield`.

Of course, feel free to ask me any questions about this.

billiaz replied:
unit tests... the worst for me :( I cannot get it yet (though pretty new with these concepts)

zuazo reviewed (changes requested):
@billiaz astonishing work! :sparkles:

I have commented some possible improvements and added some notes to help you complete the unit tests.

billiaz commented on recipes/create_pwfile.rb:
```diff
@@ -0,0 +1,59 @@
+  variables(
+    credentials: credentials
+  )
+  only_if { update_credentials }
```
it is added on my next commit ;)

billiaz commented:
Thanks for the help! it is my first real contribution.

varet80 committed:
Align with Mantainer's requests

- Update recipes with more efficient code
- Use the awesome Unit test made by the Maintainer

billiaz commented:
Is there something more I should do to make this pull request ready for merge?

zuazo commented:
Thanks @billiaz. No, it is ready to be merged. Give me a couple of days because I want to test some of the functions by hand. I'm planning to release this during weekend if you can wait a little.

zuazo commented:
OK. I have merged this PR after:

- Adding some unit tests for the new library.
- Documented the new functions.
- Changed the code a little to avoid disabling RuboCop's `LineLength`.

You can see the changes in 667058b1ccb2f52ac432cca61bbaadb58800e6a6.
